build(Makefile): make python3-config protable

tai271828 opened the pull request:
When using python virtual environment, python-config or python3-config are maybe not included in PATH, and thus the command may return incorrect information of python configuration.

tai271828 committed:
build(Makefile): make python3-config protable

When using python virtual environment, python-config or python3-config
are maybe not included in PATH, and thus the command may return
incorrect information of python configuration.

yungyuc commented on cpp/binary/pymod_modmesh/CMakeLists.txt:
```diff
@@ -34,7 +34,7 @@ if (CLANG_TIDY_EXE AND USE_CLANG_TIDY)
 
 execute_process(
-    COMMAND python3-config --extension-suffix
+    COMMAND $(dirname $(realpath $(which python)))/python3-config --extension-suffix
```
I'd be curious whether or not windows accepts it.

tai271828 replied:
I am wondering too. I am thinking if you know someone interested in testing this?

yungyuc replied:
It actually passes CI ... well, interesting.

But it does not hurt to use cmake native helpers: https://cmake.org/cmake/help/v3.0/command/get_filename_component.html

tai271828 replied:
Thanks for the tip. Let me check.

yungyuc commented on Makefile:
```diff
@@ -37,7 +37,11 @@ CMAKE_ARGS ?=
 FORCE_CLANG_FORMAT ?=
 
-pyextsuffix := $(shell python3-config --extension-suffix)
+WHICH_PYTHON := $(shell which python)
```
It mush be `python3`.  We should not allow any possibility of running python 2.

tai271828 replied:
This is fair. I will update it.

yungyuc commented on Makefile:
```diff
@@ -37,7 +37,11 @@ CMAKE_ARGS ?=
 
-pyextsuffix := $(shell python3-config --extension-suffix)
+WHICH_PYTHON := $(shell which python)
+REALPATH_PYTHON := $(shell realpath $(WHICH_PYTHON))
```
Please use make function while we can: https://www.gnu.org/software/make/manual/html_node/File-Name-Functions.html .  It runs faster than shell.

tai271828 replied:
Let me check. Thanks for the tip!

tai271828 commented on Makefile:
```diff
@@ -37,7 +37,11 @@ CMAKE_ARGS ?=
+REALPATH_PYTHON := $(shell realpath $(WHICH_PYTHON))
+DIRNAME_PYTHON := $(shell dirname $(REALPATH_PYTHON))
+
+pyextsuffix := $(shell $(DIRNAME_PYTHON)/python3-config --extension-suffix)
```
FWIW, this will bring the build flow the support of using python virtual environment.

For example, in my case it is:
- `which python`: /home/tai271828/.cache/pypoetry/virtualenvs/modmesh-venv-py3-9-13-LgJ7FNfF-py3.9/bin/python
- `realpath`: /home/tai271828/.pyenv/versions/3.9.13/bin/python3.9
- `dirname`: /home/tai271828/.pyenv/versions/3.9.13/bin/

and additionally:
```
$ ls -alh /home/tai271828/.cache/pypoetry/virtualenvs/modmesh-venv-py3-9-13-LgJ7FNfF-py3.9/bin/python
lrwxrwxrwx 1 tai271828 tai271828 49 Dec  9 21:46 /home/tai271828/.cache/pypoetry/virtualenvs/modmesh-venv-py3-9-13-LgJ7FNfF-py3.9/bin/python -> /home/tai271828/.pyenv/versions/3.9.13/bin/python
```

yungyuc replied:
Please summarize in a comment line that the treatment is for virtual env.  It should also be mentioned that a virtual env is strongly discouraged in developing modmesh.  We do not actively resolve bugs related to any virtual env including venv or conda.

tai271828 replied:
Will do!

yungyuc commented:
I would encourage you to make sure CI passes in your fork before filing a PR.

tai271828 committed:
build(Makefile): use Python3 explicitly

We should not allow any possibility of running Python 2.

tai271828 commented:
> I would encourage you to make sure CI passes in your fork before filing a PR.

This is a good idea. I will do that next time.

tai271828 committed:
test(pytest): use cmake native helper

By using cmake native helpers, e.g. get_filename_component, we can make
sure the job able to run across platforms.

tai271828 committed:
build(Makefile): use make functions

using make functions while we can since it runs faster than shell

tai271828 committed:
build(comment): add more comments for pull request 177

https://github.com/solvcon/modmesh/pull/177

tai271828 commented:
Updated the pull request according to the above comments and suggestions. Ready to review again. Thank you!

yungyuc reviewed (approved):
LGTM.  Thanks a lot.
